File: test_simple_3d.py
import os.path
import numpy as np
import matplotlib.pyplot as plt
import solvers
import lib_misc

import model_simple_3d as m
import scipy.linalg as la

# Make simulation deterministic
np.random.seed(0)

plotter = m.MainModesPlotter(m.ip, show_weights=True, cutoff=10000,
                             contours=False)
# plotter = m.AllCoeffsPlotter(m.ip, show_weights=True)
# plotter = m.MainModesPlotter(m.ip, show_weights=True)


# PRECONDITIONING {{{1

precond_vec = np.zeros(m.ip.d)
precond_mat = np.eye(m.ip.d)

preconditioning = True
if preconditioning:

    precond_vec_file = "data_julia/precond_vec.txt"
    precond_mat_file = "data_julia/precond_mat.txt"

    if os.path.exists(precond_vec_file):
        precond_vec = np.loadtxt(precond_vec_file)
        precond_mat = np.loadtxt(precond_mat_file)

precond_mat = la.sqrtm(precond_mat)

# MULTISCALE METHOD {{{1
# Test MD solver
solver_md = solvers.MdSolver(
    J=8,
    delta=1e-5,
    sigma=1e-5,
    dt=(1 if preconditioning else 1/m.k**4),
    reg=False,
    noise=False,
    parallel=True,
    adaptive=False,
    dt_min=1e-7,
    dt_max=.1,
    precond_vec=precond_vec,
    precond_mat=precond_mat,
    dirname=m.__name__ + ("-precond" if preconditioning else "-noprecond"))

# Initial parameters
theta = np.zeros(m.ip.d)

simulation = solvers.MdSimulation(
        ip=m.ip,
        initial=theta,
        solver=solver_md)

n_iter = (200 if preconditioning else 4000)
for i in range(n_iter):
    if i % 100 == 0:
        plotter.plot(i, simulation.get_data())
        plt.pause(1)
        plt.draw()

    data = simulation.step()
    print("Reg least squares {}".format(data.value_func))

File: test_simple_3d.jl
import LinearAlgebra
import Random
import Statistics
import DelimitedFiles
import DelimitedFiles
Random.seed!(0)

# Shorthand names for modules
la = LinearAlgebra
dlm = DelimitedFiles

include("lib_inverse_problem.jl")
include("solver_multiscale.jl")
include("solver_aldi.jl")
include("model_simple_3d.jl")

model = ModelSimple3d

J = 5
dtmax = 1
reg = false
opti = false
adaptive = false
config = Aldi.Config(.001, dtmax, reg , opti,  adaptive)

# Initial ensembles
all_ensembles = zeros(model.ip.d, 0)
ensembles = Random.randn(model.ip.d, J)

for iter in 1:1e4
    ensembles = Aldi.step(model.ip, config, ensembles);
end

mean = Statistics.mean(ensembles, dims=2)
cov = Statistics.cov(ensembles, dims=2)
config = Aldi.Config(.01, dtmax, reg , opti,  adaptive)

for iter in 1:9e4
    ensembles = Aldi.step(model.ip, config, ensembles);
    mean = (iter*mean + Statistics.mean(ensembles, dims=2))/(iter + 1)
    cov = (iter*cov + Statistics.cov(ensembles, dims=2))/(iter + 1)
    if iter % 1000 == 0
        println("Iter = $iter Mean = $(mean[1]), Cov = $(cov[1,1])")
    end
end

DelimitedFiles.writedlm("data_julia/precond_mat.txt", cov);
DelimitedFiles.writedlm("data_julia/precond_vec.txt", mean);
